Feature/2824 date input

kwongz opened the pull request:
### Description

Created DateInput component that builds off the DateRange structure. Redirects /components/date-range/ --> /components/date-input/

Allows users to access a single date or ranged date input calendar

![image](https://github.com/user-attachments/assets/2e244151-3d1c-4a4f-b38a-5a488f63ff19)

![image](https://github.com/user-attachments/assets/ab91490c-7be1-4382-ae92-0d0934602abc)


### Checklist

- [x] For UI or styling changes, I have added a screenshot or gif showing before & after
- [x] I have added a [changeset](https://github.com/evidence-dev/evidence/blob/main/CONTRIBUTING.md#adding-a-changeset)
- [ ] I have added to the docs where applicable
- [ ] I have added to the [VS Code extension](https://github.com/evidence-dev/evidence-vscode) where applicable

kwongz commented:
fixes #2824

archiewood commented on sites/docs/pages/components/date-input/index.md:
```diff
@@ -0,0 +1,302 @@
+    data={orders_by_day}
+    dates=day
+    title="Select a Date Input"
+    range
```
this example uses a range, but is in the wrong section? remove?
```suggestion
```

archiewood commented on sites/docs/pages/components/date-input/index.md:
```diff
@@ -0,0 +1,302 @@
+select 
+    *
+from ${orders_by_day}
+where day = '${inputs.range_filtering_a_query.value}'
```
This example chart looks odd, A barchart with a single day is rarely used.

I think you'd be using this as a one sided filter. 

```suggestion
where day > '${inputs.range_filtering_a_query.value}'
```

kwongz replied:
oh nice yeah, i like this better

archiewood commented on sites/docs/pages/components/date-input/index.md:
```diff
@@ -0,0 +1,302 @@
+/>
+<PropListing 
+    name="title"
+    options="string"
```
I think it looks odd to just have a floating number in a button. not obvious what to do without a title?

Maybe we can make this title a default?

I know @mcrascal has an different view so happy to discuss

kwongz replied:
yeah it's definitely missing something, and easily overlooked on its own

archiewood commented on sites/docs/pages/components/date-input/index.md:
```diff
@@ -0,0 +1,302 @@
+---
+title: Date Input
+sidebar_position: 1
```
should have a description

archiewood commented on sites/docs/pages/components/date-input/index.md:
```diff
@@ -0,0 +1,302 @@
+- orders_by_day.sql
+---
+
+Creates a date picker that can be used to filter a query.
```
```suggestion
A date input component allows the user to select a date or a range of dates. The selected dates can be used as inputs to queries or components.
```

archiewood replied:
This can be the description also

kwongz commented:
@archiewood   @mcrascal 

Date Input w/ calendar

![image](https://github.com/user-attachments/assets/680a4f4c-1cf6-4b40-b463-1bb74f458136)

Date Input w/ inline title

![image](https://github.com/user-attachments/assets/8ec311c5-6207-4df7-a876-e847e0f231d6)

Date Input w/ range + title (no calendar icon)

![image](https://github.com/user-attachments/assets/6fe78876-10d4-47a3-98bb-965056b35637)

archiewood commented:
The separating bar between the Date Input and Jan 1, 1970 looks different to that used in the Dropdown to me (too heavy, does not extend high or low enough)

Can we match the styling

kwongz commented:
![image](https://github.com/user-attachments/assets/8a417c84-b2d2-4c1a-a323-53d7445c33e7)

![image](https://github.com/user-attachments/assets/b7895842-d847-4148-9c63-5d6af1aa99a2)

kwongz commented:
i think we may also want to consider darkening the color of the calendar icon to match the text

before:
![image](https://github.com/user-attachments/assets/af897fb6-ed40-48e3-a033-4680fc7f68ac)

after:
![image](https://github.com/user-attachments/assets/045c50f3-9271-4227-a514-e6a42d8e95eb)

dropdown styles:
![image](https://github.com/user-attachments/assets/f71a988a-9321-476f-8001-06b313a898ef)

archiewood commented:
I find that a bit dominating at the moment, hence the suggestion

Are there any other calendar icons in simpleIcons / heroicons / whatever the other icon pack we sometimes use is that are simpler?

kwongz commented:
I think this is my preferred styling for the `calendar` icon form `tabler-icons`, just did some polishing for sizing and aligning

- default font & stroke styling
![image](https://github.com/user-attachments/assets/65f3b971-bc0c-42b7-a387-d2cccc467513)

_For Comparison:_

1. text-font-500
![image](https://github.com/user-attachments/assets/5e129262-66c9-4a0e-b215-8ebcd0b1ec4a)
2. stroke-1
![image](https://github.com/user-attachments/assets/cb29deef-7023-4f99-88db-54c669fac5d7)

kwongz commented:
lowered the stroke and change text coloring

![image](https://github.com/user-attachments/assets/93d81f4b-5ad3-4859-8cf8-4f3650b11b8b)

archiewood commented:
I think maybe `stroke-[1.8px] text-gray-700` with `Calendar Event` might look good?

archiewood commented:
although I actually like what is currently in the deploy preview

kwongz commented:
@archiewood 
I just uploaded a preview of the calendar event, i think they both look good, I'm leaning towards Calendar Event now haha

![image](https://github.com/user-attachments/assets/7ca7f44b-fdb1-4b25-a15d-4803db750969)

![image](https://github.com/user-attachments/assets/4e5be9de-d296-497c-aab4-a5fa12112197)

archiewood commented:
Calendar event looks great!

archiewood commented on packages/ui/core-components/src/lib/atoms/inputs/date-range/DateRange.svelte:
```diff
@@ -39,6 +39,9 @@
 	export let defaultValue;
 
+	/** @type {boolean} */
+	let range = true;
```
Make sure we can accept the string 

```
<DateInput
  range=true
/>
```

not just 

range={true}

archiewood replied:
See other components like 

BarChart labels=true for examples

csjh replied:
note the preferred syntax should just be `<DateInput range />` (which svelte turns into `<DateInput range={true} />`

kwongz replied:
@csjh Thanks, Sounds good. Docs are set as <DateInput range /> for the examples.
@archiewood added the ToBoolean check for the range prop to accept strings and booleans

archiewood reviewed (approved):
This also needs engineering review as the changes are quite surgical.

kwongz commented:
@ItsMeBrianD 

Moved some functions used both in DateRange and DateInput, into a helper.js 
Removed _DateRange (not being used)

I tried moving some of the query-building logic/ input store, but it felt like my changes were making the code more complicated, so I left them in the respective components.

csjh commented on packages/ui/core-components/src/lib/atoms/inputs/date-input/DateInput.svelte:
```diff
@@ -0,0 +1,128 @@
+
+	$: range = toBoolean(range);
+
+	const exec = getQueryFunction();
```
@ItsMeBrianD is there an alternative we should prefer to direct `Query.create`/`getQueryFunction` here

archiewood commented on packages/ui/core-components/src/lib/atoms/inputs/date-input/DateInput.svelte:
```diff
@@ -0,0 +1,128 @@
+		if (range) {
+			$inputs[name] = { value: undefined, start: startString, end: endString };
+		} else {
+			$inputs[name] = { value: startString, start: startString, end: undefined };
```
My thought was start is probably worth keeping as someone might change from dateinput range=true to range=false, and then it break the rest of their code

but I'm not opposed to removing it if you think this is confusing behaviour

kwongz replied:
Sounds good i'll omit the unused/duplicate value props

archiewood commented on packages/ui/core-components/src/lib/atoms/inputs/date-input/DateInput.svelte:
```diff
@@ -0,0 +1,128 @@
+	/** @type {string | undefined} */
+	export let defaultValue;
+	/** @type {boolean} */
+	export let range = false;
```
this should accept a string also
```
<DateInput
  range=true
/>
```

as well as 

<DateInput
  range={true}
/>
```

kwongz replied:
added string type to range prop

kwongz commented on packages/ui/core-components/src/lib/atoms/inputs/date-input/_DateInput.svelte:
```diff
@@ -176,13 +183,17 @@
-		selectedDateRange = targetPreset.range ?? targetPreset.value;
 		selectedPreset = targetPreset;
+		if (range) {
+			selectedDateInput = targetPreset.range ?? targetPreset.value;
```
that is accurate, we removed the presets for `dateinput range=false`, thanks!

kwongz commented on packages/ui/core-components/src/lib/atoms/inputs/date-range/DateRange.svelte:
```diff
@@ -39,6 +36,9 @@
 	export let defaultValue;
 
+	/** @type {boolean} */
+	let range = true;
```
removed prop, is something like this better

![image](https://github.com/user-attachments/assets/f1f7d456-fba1-4387-9ad4-fd6e53afa3d8)

zachstence replied:
Yes, remove the `range` variable entirely.

You can also use the shorthand for boolean props:

```html
				<DateInput
					bind:selectedDateInput
					start={startString}
					end={endString}
					loaded={loaded?.ready ?? true}
					{presetRanges}
					{defaultValue}
					range <!-- this is equivalent to range={true} -->
				/>
```

kwongz commented:
To dos:

Take DateRange themes styling and apply them to _Date Input

zachstence commented on packages/ui/core-components/src/lib/atoms/inputs/date-input/_DateInput.svelte:
```diff
@@ -184,13 +191,16 @@
 		if (!targetPreset) return;
-		selectedDateRange = targetPreset.range ?? targetPreset.value;
 		selectedPreset = targetPreset;
+		console.log(targetPreset);
```
Remove debug log

```suggestion
```

csjh reviewed (approved):
aside from this small fix and the `buildReactiveInputQuery` thing everything looks fine to me

ItsMeBrianD commented on packages/ui/core-components/src/lib/atoms/inputs/date-input/DateInput.svelte:
```diff
@@ -0,0 +1,121 @@
+
+<script>
+	import DateInput from './_DateInput.svelte';
+	// import { Query } from '@evidence-dev/sdk/usql';
```
```suggestion
```
